### ISSUE TITLE
cb() never called

### ISSUE THREAD

**Issue by buckleyGI** [microsoft/azure-pipelines-tasks#13971]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: Npm

## Environment
- Server - Azure Pipelines
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 

https://dev.azure.com/selligentsim/ES/_build/results?buildId=164920&view=results
buildId=164920
definitionId=552

- Agent - Hosted: 
    
    - If using Hosted agent, provide agent queue name:

Agent name: 'Hosted Windows 2019 with VS2019 3'
Agent machine name: 'WIN-Q1IKC5086MU'
Current agent version: '2.177.1'

## Issue Description

After doing more than 500 runs on our buildserver (Azure DevOps) we come to the the following conclusions.

- If the registry is Azure DevOps it fails in **12% of the cases** with "error cb() never called!"
(There is a module called Azure artifacts which caches the npm registry)

- if the registry is npm (https://registry.npmjs.org) **it succeeds 100% of the time.**

1 thing to note is that the "npm ci" command we are performing takes about 3 minutes when connecting to npm and 5 mins when connecting to DevOps.

Our setup:npm/6.14.8 node/v14.15.0 win32 x64
We tried also with the latest npm v7 which gives the same error.
Looking at the logs it happens 100% of the time after ckeditor-4.12.1 is handled.

For the moment we switched to the npm registry but we expect to be throttled at some point which was to reason to start relying on Azure Artifacts.

This issue seems elusive and quit a few people are struggling with it. You can count on us to provide any addition information.

I know that most issues report under the "cb() never called" cause are routed to the npm project but in this case I find it striking that it only fails on DevOps. (As stated we ran the tests 500 times so this is no fluke).

### Task logs

Can I sent them in private as they may contain secrets.
I'll send the logs of a run with system.debug=true

### Error logs

error cb() never called!
error This is an error with npm itself. Please report this error at: https://npm.community

**Comment by PrestonR** [comment]
Can confirm this issue is still happening as described above with the exact same setup.

**Comment by djentelmenis** [comment]
I feel I have the same problem in our project - ever since we added a self hosted npm package from Azure Arftifacts

**Comment by buckleyGI** [comment]
@djentelmenis We have a ticket open with MS which we already spent a lot of time on. It is currently closed with the status "not enough info" (which is not justified IMO). Maybe you can add your conclusions to there as well? We still have the problem and are able to reproduce.

https://developercommunity.visualstudio.com/content/problem/1268824/cb-never-called.html

**Comment by vaindil** [comment]
We are also running into this issue, we have to restart builds all the time because they fail with this exact error. It's incredibly irritating.

**Comment by hvdb** [comment]
We also have this problem.
We have to restart it several times until it works, very irritating and also costs lost of time..

Using ubuntu 18-04 and 20-04 and running NPM CI.

**Comment by djentelmenis** [comment]
Update from my side - Last week I tested out switching from `npm` to `yarn` and have had no more problems since.

**Comment by aldodfm** [comment]
Same here,
what we rollback npm task to:
- task: Npm@1.179.1
inputs:
command: 'install'
customRegistry: 'useFeed'
customFeed: 'XXXXXXXXXXXXXXXXXX'
Then, it doesn't fail...

**Comment by erjicles** [comment]
We are experiencing the same issue. We have a custom .npmrc file that points to our own private feed in Azure Artifacts. The npm install task fails randomly ~20% of the time with this error.

**Comment by leepowelldev** [comment]
Same issue on npm v6.

Completely different issues using npm v7 where we are getting 401 errors using Azure Artifacts as our registry and the automatically generated access tokens only work on *.visualstudio.com and not on dev.azure.com

**Comment by WolfyUK** [comment]
For me this was an expired PAT. Renewed and now all is good.

**Comment by dfbaskin** [comment]
It's definitely an intermittent error.  Also, I can't seem to make it fail if I "enable system diagnostics".

**Comment by dfbaskin** [comment]
Added a step to upgrade npm to the latest (v7.10.0) to see if that has an effect.

```
      - task: Npm@1
        displayName: "npm upgrade to latest"
        inputs:
          command: custom
          customCommand: install -g npm@latest
```

**Comment by dfbaskin** [comment]
The build is still failing, but for a different reason now (after the upgrade of npm):

```
2021-04-28T18:07:50.3096817Z [command]C:\Windows\system32\cmd.exe /D /S /C "C:\hostedtoolcache\windows\node\14.16.1\x64\npm.cmd --version"
2021-04-28T18:07:51.3602386Z 7.11.1
2021-04-28T18:07:54.8678861Z registry = "https://pkgs.dev.azure.com/.../_packaging/cai/npm/registry/" 
2021-04-28T18:07:54.8686612Z [command]C:\Windows\system32\cmd.exe /D /S /C "C:\hostedtoolcache\windows\node\14.16.1\x64\npm.cmd install"
2021-04-28T18:10:32.7797232Z npm ERR! code ERR_SOCKET_TIMEOUT
2021-04-28T18:10:32.7798425Z npm ERR! errno ERR_SOCKET_TIMEOUT
2021-04-28T18:10:32.7799541Z npm ERR! request to https://pkgs.dev.azure.com/.../_packaging/cai/npm/registry/import-local/-/import-local-2.0.0.tgz failed, reason: Socket timeout
```

**Comment by sander1095** [comment]
I can try using npm7 at some point as some people suggest this helps, but another approach where I need some help with is the following.

I use an .npmrc with the official npm repo, and my private repo with 1 private package. 
Locally i use that vsts-auth package to make this work, but in azure pipelines I dont use my npmrc but point towards my private feed which has npm as an upstream source. I do this because I couldnt get the .npmrc to work in my pipeline because of auth issues; my ipeline couldnt reach the private feed. I use hosted ubuntu lts. It basically is "use node, npm ci, npm run build". No other special stuff.

Correct me if I am wrong, but I think I would prefer using 2 package sources instead of only using my own and making npm an upstream source. This might fix this issue and would also allow me to use npm audit again because that doesnt work on private package feeds.

What do you guys think of this approach? If you guys test this, do you also have auth issues? The only downside I could come up with is that if npm is down, I can't restore packages. With an upstream source I would be able to install because it would serve as a back-up, but I dont see this as a real issue.

**Comment by vaindil** [comment]
I haven't been able to get a PR through today because our build gate (which uses `npm ci`) has been failing for hours with this error. Normally we can re-queue once or twice and it'll go through, but I'm on my 11th try in a row now and it's just not working. We did try switching to npm 7, but that doesn't work per npm/cli#2619. This is an issue that _needs_ to be addressed.

**Comment by sander1095** [comment]
@robingenz Thanks. Just tried this... but still fails..

I might have some kind of fix, but it's nothing magical. I verify the cache when the `npm ci` fails and then try again.
I encountered `npm ci` failure once because of `cb() never called`, and then after `npm cache verify` `npm ci` worked again.

The problem with this approach is that all the next steps will not run because `succeeded()` is false. So I would need to add a custom condition to every task... very cumbersome. :/ 

This is probably not a true fix. It might just be a coincidence 🤷 .
Our situation:

- Our private package source has 1 npm package that is private.
- Upstream sources are configured so we also have the npm packages we use publicly in there too.

```
    - task: Npm@1
      displayName: 'Install Dependencies'
      inputs:
        command: 'ci'
        workingDir: '$(frontendLocation)'
        verbose: false
        customRegistry: 'useFeed'
        customFeed: '7938f311-0976-4bf9-a7d5-9ccbab40c95b/7d12203e-bf6e-4128-86d4-0dc8da7b3581'
    
    - bash: echo '##vso[task.setvariable variable=npm-ci-failed]true'
      displayName: 'Set up cb() never called error workaround'
      condition: failed()
      failOnStderr: true
    
    - task: Npm@1
      displayName: 'Verify npm cache for cb() never called workaround'
      inputs:
        command: 'custom'
        workingDir: '$(frontendLocation)'
        customCommand: 'cache verify'
        customRegistry: 'useFeed'
        customFeed: '7938f311-0976-4bf9-a7d5-9ccbab40c95b/7d12203e-bf6e-4128-86d4-0dc8da7b3581'
      condition: eq(variables['npm-ci-failed'], true)
      
    - task: Npm@1
      displayName: 'Install Dependencies again (cb() never called workaround)'
      inputs:
        command: 'ci'
        workingDir: '$(frontendLocation)'
        verbose: false
        customRegistry: 'useFeed'
        customFeed: '7938f311-0976-4bf9-a7d5-9ccbab40c95b/7d12203e-bf6e-4128-86d4-0dc8da7b3581'
      condition: eq(variables['npm-ci-failed'], true)
```

**Comment by sander1095** [comment]
I am tagging the following people to get some word from microsoft on this issue.
These people have made commits on this repo around this year and are a member of the Microsoft organization on Github, so perhaps they can be of help.

@damccorm @hiyadav  @ds-ms  @arjgupta @stephenmichaelf  

This issue makes using Azure Pipelines very very "stressful". Everytime I do a build I need to keep an eye on the result of the build because I might need to retry it once or sometimes even twice because of this issue.

**Comment by damccorm** [comment]
@anatolybolshakov could your team please prioritize taking a look at this one?

**Comment by alexander-smolyakov** [comment]
@damccorm sure, let us take a look at this issue.

**Comment by anatolybolshakov** [comment]
Hi @sander1095 could you please share the pipeline logs containing this error to take a look? You can send them to my email: v-anbols@microsoft.com

**Comment by anatolybolshakov** [comment]
This also looks similar to https://github.com/npm/cli/issues/417
@sander1095 have you tried to use 7.x npm version also as described in [ticket](https://developercommunity.visualstudio.com/t/cb-never-called/1268824)?

**Comment by vaindil** [comment]
We tried to use npm 7.x with our pipelines, but 7.x doesn't seem to work with DevOps, see npm/cli#2619. I spent several hours troubleshooting and trying various things to get it to work before we had to just downgrade our agents back to 6.x.

**Comment by sander1095** [comment]
@anatolybolshakov  I have sent the logs your way.

I believe I did try npm 7 for a while, but I can't remember if that really fixed it, sorry. In any case, using npm 7 is not really a good fix for me, we use angular 11/12 which uses node 14 which uses npm 6. I believe node 16 will start having npm 7 included by default; that is when we start using npm 7 :).

Because I will be on holiday for 2 weeks, see below for a step-by-step guide on how to reproduce this issue. If this doesn't work, I could set up a a public repo when I am back so it is easier to look into.

Node 14.17.3 x64
NPM 6.14.13

- Create a Azure DevOps project
- Create a new Azure Repo
- Create a Azure Artifacts feed
- Host an NPM package in there. Perhaps just create one yourself and put it in there.
- Now, I would advice to create an Angular 11 project. This is probably not required, but it is the easiest way I can think of to get close to my situation.
- Now, connect to your feed. Azure Artifacts has a "Connect to feed" button that will bring you to a page explaining how you connect to your feed.
- Install your private package in your project.
  - In our case, some packages in our package lock json still reference the public npm registry while others reference our private package feed. Our private package feed also contains the packages from npm registry as an upstream source(?). This doesn't really matter though, deleting the package lock and reinstalling everything to ensure that ALL packages reference the private source will still result in the `cb() never called` error.


You are now fully setup with a repo that works locally but on Azure Devops will fail.

I would recommend the following YAML setup:

```
name: $(major).$(minor)$(rev:.r)

pool:
  vmImage: 'ubuntu-latest'

stages:
- stage: Build
  displayName: Build
  jobs:
  - job: Build_Code
    displayName: Build code
    steps:
    - task: NodeTool@0
      displayName: 'Use Node 14.x.x'
      inputs:
        versionSpec: 14.x.x
  
    - task: Npm@1
      displayName: 'Install Dependencies'
      inputs:
        command: 'ci'
        workingDir: 'PUT_YOUR_WORKING_DIR_HERE'
        verbose: false
        customRegistry: 'useFeed'
        customFeed: 'SELECT_YOUR_FEED_HERE'
```

Notes:

- As people in this thread already mentioned, sometimes it works, sometimes it won't.
- You could use `npm authenticate` so you can use the  `.npmrc`  instead of a custom feed, but this doesn't make a difference.

**Comment by anatolybolshakov** [comment]
@sander1095 thanks! Per this [doc](https://github.com/npm/cli/wiki/%22cb()-never-called%3F-Exit-handler-never-called%3F-I'm-having-the-same-problem!%22) looks like this error could happen for variety of reasons. 
I also tried to reproduce it, but it works fine for my configuration - could you please share some public repo to check when you're able to?
It would be also helpful to try it with npm 7 and see if it's still reproducible - to determine if it's some npm version specific issue or not.

**Comment by sander1095** [comment]
Hello! 

I will be back from my holiday in 2.5 weeks. Then I shall find some time to try out these things :) until then I will ask upon other people in this thread to give it a shot!

**Comment by anatolybolshakov** [comment]
Hey @sander1095 just checking for any updates, thanks!

**Comment by sander1095** [comment]
Hello! I have been at work again for 2 days. In the next few days or next week I want to set a reproducable up for you!

**Comment by sander1095** [comment]
I spent some time yesterday to create a repo to reproduce the problem. I have not been able to reproduce it :(. 

Perhaps I could enable more `verbose` logging and system diagnostics and send it to you @anatolybolshakov ?
Furthermore, the creator of this issue (@buckleyGI ) says in the issue they also had logs with sys diagnostics, perhaps that could be of help? I don't have access to these, unfortunately.

Otherwise I will ask other people in this thread to create some reproducable repo's. :) 

My repo's setup is basically as follows:

- private project with private package source with upstream npm. 
- upload your own package to that private package source. I just created a package that console logs something and used an azure pipeline with npm install, npm run build and the npm deploy task.
- install that package in your project's package.json, which is also in that private project
- Create an azure pipelines for the project with an `npm ci` task, using the Hosted Agent (ubuntu, but it doesn't seem to matter) . The task fails with cb() never called.

If I have more time this week I will continue digging around!

**Comment by anatolybolshakov** [comment]
@sander1095 yep, it would be great to check any verbose logs, thanks 👍

**Comment by anatolybolshakov** [comment]
@sander1095 just checked verbose npm logs, but have not found anything relevant...

> I update the NPM version to the latest in my pipeline, which currently is 7.21.1.

This could mean that there was something on npm side being fixed with later version - could you please create ticket on https://npm.community for npm team to take a look also?

**Comment by sander1095** [comment]
As I understand it, they do not fix bugs that are v6 specific, unless you can also reproduce it on v7, which I can't until now. 

I suppose that if Microsoft/a Microsoft Employee asks them about this, they might be willing to look further into it?


https://github.com/npm/cli/issues/3351

**Comment by anatolybolshakov** [comment]
@sander1095 looks like npm v6 is out of support at the moment - I would suggest to use npm v7 then since this issue seems to be already resolved for this version.
Since this issue comes from npm itself - not sure if there is anything we could do on DevOps side to resolve this issue...

**Comment by sander1095** [comment]
If there truly is no way to fix this with npm6 (which would be a shame for reasons mentioned before) a lot of documentation could be provided. If someone uses npm6 with a private registry, a warning could be logged during install or a tooltip could be provided on the task. The microsoft docs could be updated, etc..

**Comment by anatolybolshakov** [comment]
@phil-hodgson could you please take a look? Do you know if we could check anything from the npm registry side?

**Comment by corradin** [comment]
@sander1095 As you already predicted, the error came back and made me very sad.
I stepped away from Azure artifacts and connect to npmjs directly for all my packages except for my organizational design system package, which I have made a scoped package. I changed my .npmrc file to only look in https://pkgs.dev.azure.com for my design system package. The rest is retrieved from npmjs.com.
You can find more on scoped packages here: https://docs.microsoft.com/en-us/azure/devops/artifacts/npm/scopes?view=azure-devops.

To summarize: I moved away from the combination private registry + upstream sources and chose scoped sources. I am aware that it may not be a solution for everyone, because of company policies, but it could help if this is not an issue for you.

**Comment by sander1095** [comment]
@corradin could you post some more info for easier reproduction? What does your npmrc exactly look like, how did you tweak the azure feed exactly, what does your azure pipeline look like?

**Comment by sander1095** [comment]
@corradin Thanks for your explanation. As I see it, the only thing that really is different is that you have disabled upstream sources in your feed? :) I don't believe the scoped package really makes a difference otherwise :).

Perhaps it would be good for your situation to consider looking into npm7 to be safe from npm downtime thanks to that upstream source thingy

**Comment by corradin** [comment]
@sander1095 I have still upstream sources enabled on the feed, but yeah I could disable it I guess. So with it still enabled it could make a difference, but I agree with you that in my case it would be simpler to just disable it.

**Comment by riupko** [comment]
I have bumped into the same issue last Monday, 80% of builds were failed. Reviewing log shows that npm/node versions were changed over the weekend from node 14.17.**5** to 14.17.**6**, npm 6.14.**14** to 6.14.**15**
As a workaround I have added step to install previous 14.17.**5** node version and it started working again. No fails as of now.
Hope it helps someone until root issue solved.
  - task: NodeTool@0
    displayName: Use Node 14.17.5
    inputs:
      versionSpec: 14.17.5

**Comment by charleaux** [comment]
> I have bumped into the same issue last Monday, 80% of builds were failed. Reviewing log shows that npm/node versions were changed over the weekend from node 14.17.**5** to 14.17.**6**, npm 6.14.**14** to 6.14.**15**
> As a workaround I have added step to install previous 14.17.**5** node version and it started working again. No fails as of now.
> Hope it helps someone until root issue solved.
> 
> * task: NodeTool@0
>   displayName: Use Node 14.17.5
>   inputs:
>   versionSpec: 14.17.5

This worked for me. Thanks!

**Comment by sethidden** [comment]
I think this is more of a "you added a private package in the middle of development" issue rather than "something's wrong with Azure Pipelines" issue

In my case, only when I turned "Verbose logging" in the pipeline task related to `npm ci` I saw the error `nan not accessible from watchpack-chokidar2:fsevents` AND cb() never called error. Before i just had cb() never called.

I removed my node_modules, removed the package-lock.json and ran npm install again. Note I'm still using npm 6 and package-lock version is still "1". The git diff was very huge, partly because **most of the packages in the package-lock.json (version 1) still had the registry.npmjs.org link instead of the azure artifacts repo link (my artifacts repo has the option that caches npm enabled)**

**Comment by anatolybolshakov** [comment]
Hi everyone, closing this issue at the moment since it's not related to Azure Pipelines side, and there are workarounds for most common related issues. Please feel free to ask any further questions.

**Comment by sander1095** [comment]
The solution proposed by @riupko sadly didn't work for us. Still have the same issue.. How is it going for you @riupko ?

@anatolybolshakov  I hate to be a bother, but what is the real workaround for npm6 users? I can't find one.. 
The only thing that might work is upgrading to npm7, but if you don't want to mess with npm versions you have to wait for node to use that version. node16 uses it, but is only in active LTS at the end of october. I think we will go that route, and hope that Angular will support node16. 

I can't really find a reason for why this isn't azure pipelines's fault, because if you use the npm registry nothing goes wrong. But perhaps next month this problem will be solved by forcing people to upgrade to npm7 or to a hopefully-compatible node version that uses npm7..

In our case we use Angular

**Comment by riupko** [comment]
@sander1095, so far I have got only one build fail this morning. It was fixed by rerunning.

**Comment by sander1095** [comment]
@riupko how many runs succeeded before one failed? On average, 10-15 runs succeed before one fails. We ran 30 builds, 4 failed. 

I am guessing you didnt have many runs yet; pinning a specific node version doesnt seem to be a fix. It could actually be harmfuk because you dont get any more security fixes in the LTS.

**Comment by riupko** [comment]
@sander1095 I have 86 builds since I have applied workaround and only 1 failed today. I will definitely look into this issue later again. It's not a fix obviously.

**Comment by jorgecf** [comment]
I can confirm this is still happening. It only happens in one projects' pipeline, but it usually takes 3-4 times before being able to complete a succesful build. It happens both with npm i and npm ci. 

We've tried the version stated by @riupko but no luck. The only way that seems to work is using npm@7, but the builds go from 3-4 minutes to something on the range of 25-35, which doesn't seem like an option.

**Comment by bensgroi** [comment]
After wrestling with this issue for months, we have mostly abandoned ADO feeds in place of the public npm registry. We've hardly had any failures since switching.

**Comment by wilanc** [comment]
We have also encountered this error and raised a support request with MS. One workaround if you don't want to upgrade npm is use the `retryCountOnTaskFailure` on the task:
```
- task: Npm@1
  inputs:
    command: 'install'
    verbose: false
    customRegistry: 'useFeed'
    customFeed: 'xxxxxxxxxxxxxxxxxxx'
  displayName: 'Npm install'
  retryCountOnTaskFailure: 5
```

**Comment by Adebeer** [comment]
@bensgroi  Thanks for the retry suggestion - nice/easy suggestion until it's fixed properly.
I've added this to our npm steps so hopefully this should get most of our builds working more reliably

**Comment by coolhome** [comment]
After trying all the mitigations people propose I think I have identified a high-level reason for why this may cause an issue.

When you downgrade node or npm the globally installed packages are not re-installed for that specific node/npm version. Microsoft Hosted Agents actually pre-installs different global NPM packages that may conflict.

Windows 2016 - NPM Globally Installed Packages
https://github.com/actions/virtual-environments/blob/06446ddb68fa8afa21cd8d2676a1757cbf6e9ce3/images/win/toolsets/toolset-2016.json#L370-L384

Windows 2019 - NPM  Globally Installed Packages
https://github.com/actions/virtual-environments/blob/e4d58452f4d4fc085b24c7be9bb6d57a3ba4d3ec/images/win/toolsets/toolset-2019.json#L406-L420

Windows 2022- NPM Globally Installed Packages
https://github.com/actions/virtual-environments/blob/main/images/win/toolsets/toolset-2022.json#L290-L298

This is why re-installed NPM after downgrading Node may solve the issue. For others they may need to re-install/uninstall other global tools like lerna, webpack, etc.

The following documentation leads you to a readme but the readme does not include the globally installed packages. **I think it would be a great idea to add this to that generated document.**
https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml
https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md
